### DISCUSSION TITLE
Add the ability to send a custom EDL authentication signature for Xiaomi devices

### DISCUSSION THREAD

**Discussion by David112x** [bkerler/edl#568] [Ideas]
I looked into how the process of sending the EDL authentication signature to a Xiaomi device works, and it works like this:

- First, an XML file is sent to the device, this XML file lets the device know that it's about to receive the signature, the XML file looks like this:
`<?xml version="1.0" ?><data><sig TargetName="sig" verbose="1" size_in_bytes="256" /></data>`
- The device replies with: 
![image](https://github.com/bkerler/edl/assets/82758916/156fb33a-4841-4dc8-894e-55f24f30a133)
- Then the signature is sent to the device, not inside an XML file no, rather the signature itself is sent directly to the device, after sending that response and "size_in_bytes" lets EDL know how much the signature is in size, so it'll know what to expect after sending that response.

The way this could work is that the user gets a hex editor and writes the signature's hex data to a file, not in plain text no, but to the file, for example:
![image](https://github.com/bkerler/edl/assets/82758916/e2f80b2d-1e28-4589-acc9-72ac8b831cfd)
The signature above is a leaked signature that *seems* to work in all Xiaomi SM8150 devices, this may also include all Xiaomi SM7150 devices (I tested it on my Poco X3 NFC and it worked) but more testing is required, as there are 3 signatures and it isn't known which of them work for specific SoCs, what is known is that one of those 3 keys work on the following SoCs:
- SD 636
- SD 680
- SD 845
- SD 855
- SD 860
- SD 870

One key may work on, say for example, SD 855, SD 845 and 636, while it may not work in the others and require a different key out of the 3.

So the way this new feature could work is that the tool is launched with the following command arguments
`edl auth --signature=<path/to/signature>`

The tool does it's usual stuff, then it sends the command for authentication, aka the XML file with "sig", waits for the device to send ACK and if the device sends ACK, then continue, but if it says NAK, error out and check if it had a "log value", if it continues, it writes the signature file directly to the device's COM port, after that, it listens in to check for response's value, if it's ACK, it means that EDL is authenticated, but it's NAK, then it means that authentication failed and that this could be due to a "signature verification error", maybe there are more errors, however, these are yet to be documented.

The reason why I think it should focus on the response value is because I don't think it would be a great idea to hardcode the string that the device sends when EDL is authenticated or when it fails, the log value's string could be different.

![image](https://github.com/bkerler/edl/assets/82758916/f2a6408d-0bd5-4084-8e80-adcc29db6714)
This is how my Poco X3 NFC responds to the signature that I took a screenshot of.

**Comment by Dinolek** [comment]
Signature here is RSA2048 encrypted, pub key is in the firehose loader.
It decrypts to:
```
0001ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff00
```

Signature currently used in this repo decrypts to:
```
0001ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff00516c4a4f5256566e5358565253544a73636a6872553164445133453164574d335a6e706f52773d3d
```

This works because token is 0(at least in my firehose), before requesting it(TargetName="req") and there is no check.
I have successfully tested it on Poco F2 Pro(sm8250).

**Reply by David112x** [reply]
It seems that it works the same way in my Poco X3 NFC as well, given that I attempted going through the entire auth process (requesting the token/blob and then sending this signature and yet it would fail.

**Reply by team-orangeBlue** [reply]
Is it time to make lookup charts for devices with broken PBL code?

**Reply by David112x** [reply]
The 3 signatures that I mentioned earlier were released with a chinese tool that took advantage of the BFF (Batch Framework for Flashing), though it had a modified version of this framework in order to be able to use EDL and authentication, either by those signatures or an user-provided signature.

**Comment by David112x** [comment]
I've been testing a few ideas for sending the custom signature over to the phone and there's a "Xiaomi" module that contains a function for attempting EDL authentication with a signature that's hardcoded into the tool, I made it so that I could enter the path to my own signature instead of hard coding it in.

[This](https://github.com/bkerler/edl/blob/master/edlclient/Library/Modules/xiaomi.py#L45) is the code responsible for checking the device's response after sending the signature if I'm not mistaken.

```
Traceback (most recent call last):
  File "/home/david/edl/./edl", line 393, in <module>
    base.run()
  File "/home/david/edl/./edl", line 384, in run
    if fh.connect(sahara):
       ^^^^^^^^^^^^^^^^^^
  File "/home/david/edl/edlclient/Library/firehose_client.py", line 114, in connect
    if self.firehose.configure(0):
       ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/david/edl/edlclient/Library/firehose.py", line 917, in configure
    if self.modules.edlauth():
       ^^^^^^^^^^^^^^^^^^^^^^
  File "/home/david/edl/edlclient/Library/Modules/init.py", line 87, in edlauth
    return self.xiaomi.edl_auth()
           ^^^^^^^^^^^^^^^^^^^^^^
  File "/home/david/edl/edlclient/Library/Modules/xiaomi.py", line 50, in edl_auth
    if "value" in rsp.resp:
       ^^^^^^^^^^^^^^^^^^^
TypeError: argument of type 'bool' is not iterable
```

That doesn't mean authentication is unsuccessful though, because it does go through and then I was able to send other commands.

**Reply by David112x** [reply]
This is a screenshot of the output of the command `edl getstorageinfo --debugmode` on my Poco X3 NFC:
![image](https://github.com/bkerler/edl/assets/82758916/77653ca8-d98a-46dc-85e0-d7ad5bbac44f)

This is how it looks like to use the same command without `--debugmode`:
![image](https://github.com/bkerler/edl/assets/82758916/768c06a2-84a6-4667-87c6-6bdff2e7f287)

I don't think it's normal for it to not output anything like this and keep in mind that I executed the "edl" that was installed in my system, not the local one in the current directory, otherwise it would've been something like `./edl`.

**Comment by David112x** [comment]
> ```
> Traceback (most recent call last):
>   File "/home/david/edl/./edl", line 393, in <module>
>     base.run()
>   File "/home/david/edl/./edl", line 384, in run
>     if fh.connect(sahara):
>        ^^^^^^^^^^^^^^^^^^
>   File "/home/david/edl/edlclient/Library/firehose_client.py", line 114, in connect
>     if self.firehose.configure(0):
>        ^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/home/david/edl/edlclient/Library/firehose.py", line 917, in configure
>     if self.modules.edlauth():
>        ^^^^^^^^^^^^^^^^^^^^^^
>   File "/home/david/edl/edlclient/Library/Modules/init.py", line 87, in edlauth
>     return self.xiaomi.edl_auth()
>            ^^^^^^^^^^^^^^^^^^^^^^
>   File "/home/david/edl/edlclient/Library/Modules/xiaomi.py", line 50, in edl_auth
>     if "value" in rsp.resp:
>        ^^^^^^^^^^^^^^^^^^^
> TypeError: argument of type 'bool' is not iterable
> ```

With the help of [commit 9c9c489: Fix xiaomi edl auth check
](https://github.com/bkerler/edl/pull/569/commits/9c9c4894376b253ade22a03e3e833d7f2662382a)

I was able to get authentication to work properly with custom signatures and the one that was hard coded into the tool

![image](https://github.com/bkerler/edl/assets/82758916/56a8535b-34a3-4117-81c1-ea4b6f3b42d7)

**Comment by team-orangeBlue** [comment]
Tested on a redmi note 10 pro. Bought in June of 2021.
Both signatures rejected
```
Qualcomm Sahara / Firehose Client V3.62 (c) B.Kerler 2018-2024.
main - Using loader /home/ubuntu/Downloads/sweet_edl_files/prog_ufs_firehose_sm7150_ddr.elf ...
main - Waiting for the device
main - Device detected :)
sahara - Protocol version: 2, Version supported: 1
main - Mode detected: sahara
sahara - 
Version 0x2
------------------------
HWID:              0x000e60e100720000 (MSM_ID:0x000e60e1,OEM_ID:0x0072,MODEL_ID:0x0000)
CPU detected:      "SM7150"
PK_HASH:           0x1bebe3863a6781db4b01086063007334de9e5ca14971c7c4f4358ec9d79cda4692ce5e948c6fd409408f4c919fcadfe3
Serial:            0x179247ff

sahara - Protocol version: 2, Version supported: 1
sahara - Uploading loader /home/ubuntu/Downloads/sweet_edl_files/prog_ufs_firehose_sm7150_ddr.elf ...
sahara - 64-Bit mode detected.
sahara - Firehose mode detected, uploading...
sahara - Loader successfully uploaded.
main - Trying to connect to firehose loader ...
firehose - INFO: ufs: SKhynix
firehose - INFO: Binary build date: Feb  5 2021 @ 22:36:20
firehose - INFO: Binary build date: Feb  5 2021 @ 22:36:20 
firehose - INFO: Chip serial num: 395462655 (0x179247ff)
firehose - INFO: Supported Functions (14):
firehose - INFO: program
firehose - INFO: read
firehose - INFO: nop
firehose - INFO: patch
firehose - INFO: configure
firehose - INFO: setbootablestoragedrive
firehose - INFO: erase
firehose - INFO: power
firehose - INFO: firmwarewrite
firehose - INFO: getstorageinfo
firehose - INFO: benchmark
firehose - INFO: emmc
firehose - INFO: ufs
firehose - INFO: fixgpt
firehose - INFO: End of supported functions 14
firehose_client
firehose_client - [LIB]: No --memory option set, we assume "UFS" as default ..., if it fails, try using "--memory" with "UFS","NAND" or "spinor" instead !
firehose - Xiaomi EDL Auth detected.
Requesting auth
Sending signature
Sent k246jlc8rQfBZ2RLYSF4Ndha1P3bfYQKK3IlQy/NoTp8GSz6l57RZRfmlwsbB99sUW/sgfaWj89//dvDl6Fiwso+XXYSSqF2nxshZLObdpMLTMZ1GffzOYd2d/ToryWChoK8v05ZOlfn4wUyaZJT4LHMXZ0NVUryvUbVbxjW5SkLpKDKwkMfnxnEwaOddmT/q0ip4RpVk4aBmDW4TfVnXnDSX9tRI+ewQP4hEI8K5tfZ0mfyycYa0FTGhJPcTTP3TQzy1Krc1DAVLbZ8IqGBrW13YWN/cMvaiEzcETNyA4N3kOaEXKWodnkwucJv2nEnJWTKNHY9NS9f5Cq3OPs4pQ==
No reply for signature
Requesting auth
Sending signature
Sent vzXWATo51hZr4Dh+a5sA/Q4JYoP4Ee3oFZSGbPZ2tBsaMupn+6tPbZDkXJRLUzAqHaMtlPMKaOHrEWZysCkgCJqpOPkUZNaSbEKpPQ6uiOVJpJwA/PmxuJ72inzSPevriMAdhQrNUqgyu4ATTEsOKnoUIuJTDBmzCeuh/34SOjTdO4Pc+s3ORfMD0TX+WImeUx4c9xVdSL/xirPl/BouhfuwFd4qPPyO5RqkU/fevEoJWGHaFjfI302c9k7EpfRUhq1z+wNpZblOHuj0B3/7VOkK8KtSvwLkmVF/t9ECiry6G5iVGEOyqMlktNlIAbr2MMYXn6b4Y3GDCkhPJ5LUkQ==
No reply for signature
Requesting auth
Sending signature
Sent k246jlc8rQfBZ2RLYSF4Ndha1P3bfYQKK3IlQy/NoTp8GSz6l57RZRfmlwsbB99sUW/sgfaWj89//dvDl6Fiwso+XXYSSqF2nxshZLObdpMLTMZ1GffzOYd2d/ToryWChoK8v05ZYAAAEQUyaZJT4LHMXZ0NVUryvUbVbxjW5SkLpKDKwkMfnxnEwaOddmT/q0ip4RpVk4aBmDW4TfVnXnDSX9tRI+ewQP4hEI8K5tfZ0mfyycYa0FTGhJPcTTP3TQzy1Krc1DAVLbZ8IqGBrW13YWN/cMvaiEzcETNyA4N3kOaEXKWodnkwucJv2nEnJWTKNHY9NS9f5Cq3OPs4pQ==
No reply for signature
Auth likely failed
firehose
firehose - [LIB]: Error on EDL Authentification```

**Reply by Dinolek** [reply]
Sweet prog_ufs_firehose_sm7150_ddr.elf has this check: `if (strlen(&token) == 0)`, so it's not going to work

**Reply by David112x** [reply]
Do you know if sweet's firehose from the engineering ROM also has that check?

**Reply by Dinolek** [reply]
eng loader has same check, device should replay with "Blob should init first"

**Reply by team-orangeBlue** [reply]
Just as I was about to say that "it failed!"..

anyway:
```
python3 edl nop --loader='/home/ubuntu/Downloads/prog_ufs_firehose_sm7150_ddr.elf'
Qualcomm Sahara / Firehose Client V3.62 (c) B.Kerler 2018-2024.
main - Using loader /home/ubuntu/Downloads/prog_ufs_firehose_sm7150_ddr.elf ...
main - Waiting for the device
main - Device detected :)
sahara - Protocol version: 2, Version supported: 1
main - Mode detected: sahara
sahara - 
Version 0x2
------------------------
HWID:              0x000e60e100720000 (MSM_ID:0x000e60e1,OEM_ID:0x0072,MODEL_ID:0x0000)
CPU detected:      "SM7150"
PK_HASH:           0x1bebe3863a6781db4b01086063007334de9e5ca14971c7c4f4358ec9d79cda4692ce5e948c6fd409408f4c919fcadfe3
Serial:            0x179247ff

sahara - Protocol version: 2, Version supported: 1
sahara - Uploading loader /home/ubuntu/Downloads/prog_ufs_firehose_sm7150_ddr.elf ...
sahara - 64-Bit mode detected.
sahara - Firehose mode detected, uploading...
sahara - Loader successfully uploaded.
main - Trying to connect to firehose loader ...
firehose - INFO: ufs: SKhynix
firehose - INFO: Binary build date: Apr 25 2021 @ 11:28:35
firehose - INFO: Binary build date: Apr 25 2021 @ 11:28:35 
firehose - INFO: Chip serial num: 395462655 (0x179247ff)
firehose - INFO: Supported Functions (15):
firehose - INFO: program
firehose - INFO: read
firehose - INFO: nop
firehose - INFO: patch
firehose - INFO: configure
firehose - INFO: setbootablestoragedrive
firehose - INFO: erase
firehose - INFO: power
firehose - INFO: quick_reset
firehose - INFO: firmwarewrite
firehose - INFO: getstorageinfo
firehose - INFO: benchmark
firehose - INFO: emmc
firehose - INFO: ufs
firehose - INFO: fixgpt
firehose - INFO: End of supported functions 15
firehose_client
firehose_client - [LIB]: No --memory option set, we assume "UFS" as default ..., if it fails, try using "--memory" with "UFS","NAND" or "spinor" instead !
firehose - Xiaomi EDL Auth detected.
firehose
firehose - [LIB]: ERROR: Only nop and sig tag can be recevied before authentication.
If you want to use your own signature, enter Y, otherwise enter N: y
Enter path to your EDL signature: /home/ubuntu/Downloads/EDL1.sig
Requested auth
Request accepted
Sent data
firehose
firehose - [LIB]: Error on EDL Authentification
ubuntu@ubuntu:~/edl$ python3 edl nop --loader='/home/ubuntu/Downloads/prog_ufs_firehose_sm7150_ddr.elf'
Qualcomm Sahara / Firehose Client V3.62 (c) B.Kerler 2018-2024.
main - Using loader /home/ubuntu/Downloads/prog_ufs_firehose_sm7150_ddr.elf ...
main - Waiting for the device
main - Device detected :)
main - Mode detected: firehose
firehose
firehose - [LIB]: ERROR: Only nop and sig tag can be recevied before authentication.
If you want to use your own signature, enter Y, otherwise enter N: n
Requested auth
Request accepted
Sent data
firehose
firehose - [LIB]: Error on EDL Authentification
```

Here's the engineering loader itself. A bit smaller. Change extension so that it's an elf.
[prog_ufs_firehose_sm7150_ddr.txt](https://github.com/user-attachments/files/16494923/prog_ufs_firehose_sm7150_ddr.txt)

So we know that the poco X3 that started with android 10 has the bug but the rn10p on a11 doesn't. Good baseline but more confirmation needed for "phones older are flawed, phones newer aren't".

**Reply by Yuki1536** [reply]
I wrote a small utility to deal with the the authentication process. The cli utility includes sending signature and acquiring the challenge code (blob). The default signature only works with devices manufacturered earlier than 2020.

**Reply by HomiRuim** [reply]
Hi @Yuki1536, can you send here the utility? I'm having problem with the blob when trying to use the official rom firehose file on the EDL tool.

**Comment by starsunyzl** [comment]
PR https://github.com/bkerler/edl/pull/569
Tested on Mi 9, not working. However, it works with some paid tools, it should be the same solution.

```
RX:<?xml version="1.0" encoding="UTF-8" ?>
RX:<data>
RX:<log value="ERROR: Blob should init first." /></data>
```

**Reply by team-orangeBlue** [reply]
Crikey...

I really wanna know how things like unlocktool approach this. Putting a wireshark and a serial logger somewhere should be easy. $25/3mo is cheap, but I need to justify my purchase in part.

**Reply by starsunyzl** [reply]
> Crikey...
> 
> I really wanna know how things like unlocktool approach this. Putting a wireshark and a serial logger somewhere should be easy. $25/3mo is cheap, but I need to justify my purchase in part.

UnlockTool has anti-logger/monitor feature, so maybe you need to do some work to bypass it. I purchased UnlockTool, but I'm not working on it right now, maybe later.

**Reply by team-orangeBlue** [reply]
I'd hotspot the host PC to another PC that has wireshark running and then sniff the wifi.

As for serial... Not exactly sure.

**Reply by David112x** [reply]
> I'd hotspot the host PC to another PC that has wireshark running and then sniff the wifi.
> 
> As for serial... Not exactly sure.

https://github.com/fjovine/SerialSniffer

**Reply by David112x** [reply]
> > Crikey...
> > I really wanna know how things like unlocktool approach this. Putting a wireshark and a serial logger somewhere should be easy. $25/3mo is cheap, but I need to justify my purchase in part.
> 
> UnlockTool has anti-logger/monitor feature, so maybe you need to do some work to bypass it. I purchased UnlockTool, but I'm not working on it right now, maybe later.

Does it have some sort of anti VM? if not then it should be workable.

**Reply by mouseos** [reply]
> > Crikey...
> > I really wanna know how things like unlocktool approach this. Putting a wireshark and a serial logger somewhere should be easy. $25/3mo is cheap, but I need to justify my purchase in part.
> 
> UnlockTool has anti-logger/monitor feature, so maybe you need to do some work to bypass it. I purchased UnlockTool, but I'm not working on it right now, maybe later.

USB over network can also be used to capture USB.
Install usb over network client on the PC with unlocktool, and install usb over network server and usbpcap on the other PC.

**Reply by team-orangeBlue** [reply]
What exactly do you suggest? The site tells you to remove flexihub, which is a USB mirroring tool.

**Comment by David112x** [comment]
Well, seeing what has happened so far, I think we can count out the Mi 9, the Mi 9T, the Redmi Note 10 Pro and the Mi 11 Lite 4G.

The Redmi Note 9S/9 Pro on the other hand is very likely to be vulnerable and the reason why I believe so is because that device and the Poco X3 NFC seem to share the same firmware to some extent, seeing how ABL/LinuxLoader has references for their codenames and such.

To me, it's kinda weird that LinuxLoader has such references on devices that are different in a lot of ways, but that's just ABL, not XBL.

**Reply by moha2508** [reply]
Pls bro how to sand auth.sig

**Comment by moha2508** [comment]
Am tasted note 10 pro and note work

**Reply by team-orangeBlue** [reply]
We already know that. The task is to figure out if there's anything that can be done to simulate things properly. The X3 fell short, after all.

**Comment by atharva288** [comment]
Iam ready to test this on redmi note 9s, I just need a Lil bit of guide on how to

**Comment by jroovy** [comment]
https://github.com/bkerler/edl/discussions/568#discussion-6888079
> as there are 3 signatures and it isn't known which of them work for specific SoCs, what is known is that one of those 3 keys work on the following SoCs

Where is the 3rd signature? I only see 2.
The 2 signatures don't work on my Poco X3 Pro.
I'd like to test the 3rd signature if available.

**Reply by jroovy** [reply]
Hi, the auth signatures are listed inside this file:
https://github.com/bkerler/edl/pull/569/commits/ec43594ccd46111cf873dba8dfca5cac90fb0e0d

**Reply by 0v3rfl00w** [reply]
Unfortunately the signatures didn't work with my poco X3 pro :(

**Reply by LrX-18** [reply]
did you tried those signed.sig, also did find a solution to recover data on poco x3 pro?

**Comment by 0v3rfl00w** [comment]
@David112x would you please share the signature files if possible , i will test with my poco x3 pro 
Thanks :)

**Comment by PTX64** [comment]
-

**Comment by Kuduxaaa** [comment]
Hello guys, Any solution? I hard-bricked my Redmi Note 10 Pro Global (sweet) by flashing incorrect ROM. Now All I have is black screen (completely off) no fastboot no recovery, only EDL mode.

firehose - [LIB]: Error on EDL Authentification

**Reply by team-orangeBlue** [reply]
it takes just a little bit of reading effort to realize that you need to pay

**Comment by id0s** [comment]
can you help for xiaomi note 11 pro 5g (veux)

**Comment by Yuki1536** [comment]
Yeah. And the signature is generated using Xiaomi's secret key.

**Comment by MatiM72737** [comment]
will it work with SD 780G?

**Comment by fmcpe** [comment]
<img width="1278" height="379" alt="image" src="https://github.com/user-attachments/assets/ad3be377-d4af-40ff-9e92-8f5bf792e868" />
Has anyone figured out a way for the X3 Pro? I’ve tried finding a way to recover it 😭

**Comment by salaheddine-neg** [comment]
in my case i have oppo a78 4G with SD680, i found a firehose that could work but m stuck in auth so what m asking if this method will work for me ?

**Comment by hdk2d** [comment]
> I looked into how the process of sending the EDL authentication signature to a Xiaomi device works, and it works like this:
> 
> * First, an XML file is sent to the device, this XML file lets the device know that it's about to receive the signature, the XML file looks like this:
>   `<?xml version="1.0" ?><data><sig TargetName="sig" verbose="1" size_in_bytes="256" /></data>`
> * The device replies with:
>   ![image](https://private-user-images.githubusercontent.com/82758916/344884780-156fb33a-4841-4dc8-894e-55f24f30a133.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzYwMjk3NTgsIm5iZiI6MTc3NjAyOTQ1OCwicGF0aCI6Ii84Mjc1ODkxNi8zNDQ4ODQ3ODAtMTU2ZmIzM2EtNDg0MS00ZGM4LTg5NGUtNTVmMjRmMzBhMTMzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjA0MTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwNDEyVDIxMzA1OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTViOGQzNjRmOTg4ZTE4MmY0NTJmNjIwNWRkYjYxMWFjZDAwZmQ0MDJkMzU4MTgwNmIxMWU3NDczMDNhYjk4NTkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JnJlc3BvbnNlLWNvbnRlbnQtdHlwZT1pbWFnZSUyRnBuZyJ9.rWnIMg8LGZd7CrCEZFVmzLRzKMxtduTllai7qXvS2rs)
> * Then the signature is sent to the device, not inside an XML file no, rather the signature itself is sent directly to the device, after sending that response and "size_in_bytes" lets EDL know how much the signature is in size, so it'll know what to expect after sending that response.
> 
> The way this could work is that the user gets a hex editor and writes the signature's hex data to a file, not in plain text no, but to the file, for example: ![image](https://private-user-images.githubusercontent.com/82758916/344869631-e2f80b2d-1e28-4589-acc9-72ac8b831cfd.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzYwMjk3NTgsIm5iZiI6MTc3NjAyOTQ1OCwicGF0aCI6Ii84Mjc1ODkxNi8zNDQ4Njk2MzEtZTJmODBiMmQtMWUyOC00NTg5LWFjYzktNzJhYzhiODMxY2ZkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjA0MTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwNDEyVDIxMzA1OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWUwZmVhOGFlMTk1ZjcyNzQ0NDRmNTc5YzA1N2YxMjg5OTMyZDZmMGJlMjRmMGFmMDFkYzhkNTJjN2RjYWNkMWMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JnJlc3BvbnNlLWNvbnRlbnQtdHlwZT1pbWFnZSUyRnBuZyJ9.9l_wOA_JzLP8alqG3kSipbQ8tOyySKRUcYygqNGtH3Y) The signature above is a leaked signature that _seems_ to work in all Xiaomi SM8150 devices, this may also include all Xiaomi SM7150 devices (I tested it on my Poco X3 NFC and it worked) but more testing is required, as there are 3 signatures and it isn't known which of them work for specific SoCs, what is known is that one of those 3 keys work on the following SoCs:
> 
> * SD 636
> * SD 680
> * SD 845
> * SD 855
> * SD 860
> * SD 870
> 
> One key may work on, say for example, SD 855, SD 845 and 636, while it may not work in the others and require a different key out of the 3.
> 
> So the way this new feature could work is that the tool is launched with the following command arguments `edl auth --signature=<path/to/signature>`
> 
> The tool does it's usual stuff, then it sends the command for authentication, aka the XML file with "sig", waits for the device to send ACK and if the device sends ACK, then continue, but if it says NAK, error out and check if it had a "log value", if it continues, it writes the signature file directly to the device's COM port, after that, it listens in to check for response's value, if it's ACK, it means that EDL is authenticated, but it's NAK, then it means that authentication failed and that this could be due to a "signature verification error", maybe there are more errors, however, these are yet to be documented.
> 
> The reason why I think it should focus on the response value is because I don't think it would be a great idea to hardcode the string that the device sends when EDL is authenticated or when it fails, the log value's string could be different.
> 
> ![image](https://private-user-images.githubusercontent.com/82758916/344876477-f2a6408d-0bd5-4084-8e80-adcc29db6714.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzYwMjk3NTgsIm5iZiI6MTc3NjAyOTQ1OCwicGF0aCI6Ii84Mjc1ODkxNi8zNDQ4NzY0NzctZjJhNjQwOGQtMGJkNS00MDg0LThlODAtYWRjYzI5ZGI2NzE0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjA0MTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwNDEyVDIxMzA1OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQyNTdmMmMzMjJmYzAxYTE1NTgxYWQ1Y2ZiODU2Njk1ZGEwMmY5NjVhOTZjMDc0NzUzOWY1MzBlYmQ4ZGIzMDcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JnJlc3BvbnNlLWNvbnRlbnQtdHlwZT1pbWFnZSUyRnBuZyJ9.19jtdwSdSjG_FD7n7mrUyIByRb9kLYc03-j0hCUrJ_s) This is how my Poco X3 NFC responds to the signature that I took a screenshot of.

Will it work with a Snapdragon 8 Gen 2 CPU on a Redmi K70?

**Reply by fmcpe** [reply]
Unfortunately, no

**Comment by cosmingsm** [comment]
Can I do this on Xiaomi mi note 10 lite (toco)?
